docs/source/en/generation_strategies.mdx

```diff
@@ -0,0 +1,304 @@
+>>> tokenizer.decode(outputs[0], skip_special_tokens=True)
+'Das Haus ist wunderbar.'
+```
+
+### Diverse beam search decoding
+
+The diverse beam search decoding strategy is an extension of the beam search strategy that allows for generating a more diverse
+set of beam sequences to choose from. To learn how it works, refer to [Diverse Beam Search: Decoding Diverse Solutions from Neural Sequence Models](https://arxiv.org/pdf/1610.02424.pdf).
+This approach has two main parameters: `num_beams` and `num_beam_groups`.
+The groups are selected to ensure they are distinct enough compared to the others, and regular beam search is used within each group.
+
+```python
+>>> from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+
+>>> checkpoint = "google/pegasus-xsum"
+>>> prompt = "The Permaculture Design Principles are a set of universal design principles \
+>>> that can be applied to any location, climate and culture, and they allow us to design \
+>>> the most efficient and sustainable human habitation and food production systems. \
+>>> Permaculture is a design system that encompasses a wide variety of disciplines, such \
+>>> as ecology, landscape design, environmental science and energy conservation, and the \
+>>> Permaculture design principles are drawn from these various disciplines. Each individual \
+>>> design principle itself embodies a complete conceptual framework based on sound \
+>>> scientific principles. When we bring all these separate  principles together, we can \
+>>> create a design system that both looks at whole systems, the parts that these systems \
+>>> consist of, and how those parts interact with each other to create a complex, dynamic, \
+>>> living system. Each design principle serves as a tool that allows us to integrate all \
+>>> the separate parts of a design, referred to as elements, into a functional, synergistic, \
+>>> whole system, where the elements harmoniously interact and work together in the most \
+>>> efficient way possible."
+
+>>> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+
+>>> model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint)
+
+>>> outputs = model.generate(input_ids, num_beams=5, num_beam_groups=5, max_new_tokens=30)
+>>> tokenizer.decode(outputs[0], skip_special_tokens=True)
+'The Design Principles are a set of universal design principles that can be applied to any location, climate and culture, and they allow us to design the most efficient and sustainable human habitation and food production systems.'
+```
+
+This guide illustrates the main parameters that enable various decoding strategies. More advanced parameters exist for the
+[`generate`] method, which gives you even further control over the [`generate`] method's behavior.
+For the complete list of the available parameters, refer to the [API documentation](./main_classes/text_generation.mdx).
+
+## Save a custom decoding strategy with your model
```

Comment:
Do you think it would make sense to move this to before "## Decoding strategies"? i.e. shape the docs to first show how to manipulate generate in general, then dive into its options.

I'm concerned users will miss this new functionality (because it's at the bottom), especially with the example removal in #21112

Reply:
Sure, folks seem to agree that your suggested order would be better. I moved it up, before the ## decoding strategies, in 50146bfd463d4d8bd94887658329b208c2f0e30f